sway/desktop/output.c

```diff
@@ -318,22 +324,93 @@ void output_damage_whole(struct sway_output *output) {
 	wlr_output_damage_add_whole(output->damage);
 }
 
+struct damage_data {
+	struct root_geometry root_geo;
+	struct sway_output *output;
+	bool whole;
+};
+
+static void damage_surface_iterator(struct wlr_surface *surface, int sx, int sy,
+		void *_data) {
+	struct damage_data *data = _data;
+	struct sway_output *output = data->output;
+	float rotation = data->root_geo.rotation;
+	bool whole = data->whole;
+
+	if (!wlr_surface_has_buffer(surface)) {
+		return;
+	}
+
+	struct wlr_box box;
+	bool intersects = get_surface_box(&data->root_geo, data->output, surface,
+		sx, sy, &box);
+	if (!intersects) {
+		return;
+	}
+
+	scale_box(&box, output->wlr_output->scale);
+
+	if (whole) {
+		wlr_box_rotated_bounds(&box, rotation, &box);
+		wlr_output_damage_add_box(output->damage, &box);
+	} else {
+		int center_x = box.x + box.width/2;
+		int center_y = box.y + box.height/2;
+
+		pixman_region32_t damage;
+		pixman_region32_init(&damage);
+		pixman_region32_copy(&damage, &surface->current->surface_damage);
+		wlr_region_scale(&damage, &damage, output->wlr_output->scale);
+		if (ceil(output->wlr_output->scale) > surface->current->scale) {
+			// When scaling up a surface, it'll become blurry so we need to
+			// expand the damage region
+			wlr_region_expand(&damage, &damage,
+				ceil(output->wlr_output->scale) - surface->current->scale);
+		}
+		pixman_region32_translate(&damage, box.x, box.y);
+		wlr_region_rotated_bounds(&damage, &damage, rotation,
+			center_x, center_y);
+		wlr_output_damage_add(output->damage, &damage);
+		pixman_region32_fini(&damage);
+	}
+}
+
 void output_damage_surface(struct sway_output *output, double ox, double oy,
 		struct wlr_surface *surface, bool whole) {
-	// TODO
-	output_damage_whole(output);
+	struct damage_data data = {
+		.output = output,
+		.whole = whole,
+	};
+
+	output_surface_for_each_surface(surface, ox, oy, &data.root_geo,
+		damage_surface_iterator, &data);
 }
 
 void output_damage_view(struct sway_output *output, struct sway_view *view,
 		bool whole) {
-	// TODO
-	output_damage_whole(output);
+	if (!sway_assert(view->swayc != NULL, "expected a view in the tree")) {
+		return;
+	}
+
+	struct damage_data data = {
+		.output = output,
+		.whole = whole,
+	};
+
+	output_view_for_each_surface(view, &data.root_geo,
+		damage_surface_iterator, &data);
 }
 
 void output_damage_whole_container(struct sway_output *output,
 		struct sway_container *con) {
-	// TODO
-	output_damage_whole(output);
+	float scale = output->wlr_output->scale;
+	struct wlr_box box = {
+		.x = con->x * scale,
+		.y = con->y * scale,
+		.width = con->width * scale,
+		.height = con->height * scale,
+	};
+	wlr_output_damage_add_box(output->damage, &box);
```

Comment:
suuuuuuuubsurfaces

Reply:
Yeah. I kind of want to restrict subsurfaces and popups to their container.

Fixed.